unittest/Dialect/TritonGPU/DialectTest.cpp

```diff
@@ -0,0 +1,126 @@
+// gtest printer for mlir::Attribute.  This must live in namespace mlir in order
+// for it to be found via ADL.
+void PrintTo(const Attribute &attr, std::ostream *os) {
+  std::string str;
+  llvm::raw_string_ostream ros(str);
+  ros << attr;
+
+  *os << str;
+}
+} // namespace mlir
+
+namespace mlir::triton::gpu {
+namespace {
+
+class InferLayoutTest : public ::testing::Test {
+public:
+  InferLayoutTest()
+      : inferLayout(
+            ctx.getOrLoadDialect<TritonGPUDialect>()
+                ->getRegisteredInterface<DialectInferLayoutInterface>()) {}
+
+protected:
+  static MLIRContext ctx;
+
+  DialectInferLayoutInterface *inferLayout;
+};
+
+/*static*/ MLIRContext InferLayoutTest::ctx;
+
+std::string strReplace(std::string s, const std::string &from,
+                       const std::string &to) {
+  size_t start_pos = 0;
+  while ((start_pos = s.find(from, start_pos)) != std::string::npos) {
+    s.replace(start_pos, from.length(), to);
+    start_pos += to.length();
+  }
+  return s;
+}
+
+// We use some abbreviations when spelling out MLIR types.
+std::string expandTyStr(std::string s) {
+  s = strReplace(s, "T<", "tensor<");
+  s = strReplace(s, "#B", "#triton_gpu.blocked");
+  s = strReplace(s, "spt", "sizePerThread");
+  s = strReplace(s, "tpw", "threadsPerWarp");
+  s = strReplace(s, "wpc", "warpsPerCTA");
+  s = strReplace(s, "ord", "order");
+  return s;
+}
+
+class InferReshapeOpNoReorderEncodingTest
+    : public InferLayoutTest,
+      public ::testing::WithParamInterface<
+          std::pair<std::string /*srcTy*/, std::string /*dstTy*/>> {};
+
+TEST_P(InferReshapeOpNoReorderEncodingTest, DoIt) {
+  std::string srcTyStr = expandTyStr(GetParam().first);
+  std::string dstTyStr = expandTyStr(GetParam().second);
+
+  auto src = mlir::parseType(srcTyStr, &ctx);
+  if (!src)
+    FAIL() << "Could not parse source type: " << srcTyStr;
+
+  auto dst = mlir::parseType(dstTyStr, &ctx);
+  if (!dst)
+    FAIL() << "Could not parse destination type: " << dstTyStr;
+
+  auto srcTy = src.cast<RankedTensorType>();
+  auto dstTy = dst.cast<RankedTensorType>();
+
+  Attribute actualDstEnc;
+  auto result = inferLayout->inferReshapeOpNoReorderEncoding(
+      srcTy.getShape(), srcTy.getEncoding(), dstTy.getShape(), actualDstEnc,
+      /*loc=*/std::nullopt);
+
+  if (dstTy.getEncoding()) {
+    ASSERT_TRUE(succeeded(result));
+    EXPECT_EQ(actualDstEnc, dstTy.getEncoding());
+  } else {
+    ASSERT_FALSE(succeeded(result));
+  }
+}
+
+// A testcase of {a, b} means, check that a reshape from shape+encoding `a` to
+// shape `b` yields encoding b.  If b has no encoding, then check that the
+// reshape is deemed impossible.
+INSTANTIATE_TEST_SUITE_P(
+    Reshapes, InferReshapeOpNoReorderEncodingTest,
+    ::testing::ValuesIn(std::vector<std::pair<std::string, std::string>>({
+        // Use raw strings in here so clang-format doesn't try to wrap them.
+        {R"(T<128x64xf32, #B<{spt=[1,1], tpw=[1,32], wpc=[1,1], ord=[1,0]}>>)",
```

Comment:
random idea while reading this: It feels like it shouldn't be very hard to write a function that maps each elements of the tensor and figure out the tuple (reg id, thread id, warp id) then runs the same thing on the destination tensor and check if things match. This way any layout can be checked without the need of a ref.
No need to change your plans, just wanted to mention it in case it helps.

Reply:
Yeah, I think I should try to make that work.  Then I can do automated testing over this as well.

Thanks for the idea.